File: python/69_Sqrt(x).py
# coding=utf-8
import unittest
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
class Solution:
    def solve(self, x):
        return self.mySqrt(x)

    def mySqrt(self, x):
        if x<=0:
            return 0
        if x<4:
            return 1
        left=1
        right=2
        while right*right<x:
            left=right
            right=right*right
        while left<=right:#注意等于符号
            mid=(left+right)/2
            if mid*mid==x:
                return mid
            elif mid*mid>x:
                right=mid-1#移位避免死循环
            else:
                left=mid+1#移位避免死循环
        return right#注意二叉查找最终状态
        
class Test(unittest.TestCase):
    #override
    def setUp(self):
        self.s=Solution()
        
    def tearDown(self):
        print "TestComplete"
        
    def genTestCase(self):
        case=[
        #-1,
        #4,
        #169,
        #100,
        #10,
        #16,
        #25,
        #33,
        48,
        49,
        50,
        ]
        self.cases=case
        
    def test(self):
        self.genTestCase()
        for case in self.cases:
            print "testcase: ", case
            res=self.s.solve(case)
            print "result: ",res
            # if assert is needed
            
unittest.main()

File: Java/69_Sqrt(x).java
public class Solution {
    public int mySqrt(int x) {
        // 二分查找，否则暴力法会超时
        if(x<=0) return x;
        else if(x<4) return 1;
        long left = 1;
        long right = 2;
        while(right*right<x){left = right; right=right<<1;}// 找到x落在的2次幂区间。这里注意，如果x很大，右边界可能溢出int，所以要用long
        long  mid=0;
        while(left<=right){
            mid = left + (right-left)/2;
            if(mid*mid==x){
                return (int)mid;
            }else if(mid*mid<x){
                left = mid+1;
            }else{
                right = mid-1;
            }
        }
        return (int)right;//~
    }
}